Add support for HPU profiler (#14773)
Lightning-AI/lightning: docs/source-pytorch/tuning/profiler_advanced.rst

4: View the profiling logs
==========================
Once the capture is finished, the page will refresh and you can browse through the insights using the **Tools** dropdown at the top left

################################################
Find bottlenecks in your code on HPU (advanced)
################################################
**Audience**: Users who want to profile their HPU models to find bottlenecks and improve performance.

----

******************
Profile HPU models
******************

> To understand the cost of each PyTorch operation, use the :class:`~pytorch_lightning.profilers.hpu.HPUProfiler` built on top of the `PyTorch profiler <https://pytorch.org/docs/1.12/profiler.html#torch-profiler>`__.

Reviewer: ```suggestion
To understand the cost of each PyTorch operation, use the :class:`~pytorch_lightning.profilers.hpu.HPUProfiler` built on top of the :class:`~pytorch_lightning.profilers.pytorch.PyTorchProfiler`.
```

.. code-block:: python

    from pytorch_lightning.profilers import HPUProfiler

    trainer = Trainer(profiler=HPUProfiler())

The profiler will dump a trace file for each profiler step, ``training_step``, ``backward``, ``validation_step``, ``test_step``, and ``predict_step`` by default.

> The user can provide ``PyTorchProfiler(record_functions={...})`` to extend the scope of profiled functions.

Reviewer: ```suggestion
The user can provide ``HPUProfiler(record_functions={...})`` to extend the scope of profiled functions.
```

    empty                  0.27%            4.456us          0.27%            4.456us          2.228us
    copy_                  0.15%            2.526us          0.15%            2.526us          2.526us
    broadcast_tensors      0.15%            2.492us          0.15%            2.492us          2.492us
    size                   0.06%            0.967us          0.06%            0.967us          0.484us
    is_complex             0.06%            0.961us          0.06%            0.961us          0.481us
    stride                 0.03%            0.517us          0.03%            0.517us          0.517us
    ---------------------  ---------------  ---------------  ---------------  ---------------  ---------------
    Self CPU time total: 1.681ms

.. note::
    Since HPUProfiler extends PyTorch Profiler, when using the HPUProfiler, wall clock time will not be representative of the true wall clock time.
    This is due to forcing profiled operations to be measured synchronously, when many HPU ops happen asynchronously.
    It is recommended to use this Profiler to find bottlenecks/breakdowns, however for end to end wall clock time use

>     the ``SimpleProfiler``.

Reviewer: ```suggestion
    the :class:`~pytorch_lightning.profilers.simple.SimpleProfiler`.
```

4: View the profiling logs
==========================
Once the capture is finished, the page will refresh and you can browse through the insights using the **Tools** dropdown at the top left

################################################
Find bottlenecks in your code on HPU (advanced)
################################################
**Audience**: Users who want to profile their HPU models to find bottlenecks and improve performance.

----

> **************************
> Profile HPU models
> **************************

Reviewer 2: ```suggestion
******************
Profile HPU models
******************
```

To understand the cost of each PyTorch operation, use the :class:`~pytorch_lightning.profilers.hpu.HPUProfiler` built on top of the `PyTorch profiler <https://pytorch.org/docs/1.12/profiler.html#torch-profiler>`__.

.. code-block:: python

    from pytorch_lightning.profilers import HPUProfiler

>     profiler = HPUProfiler()
>     trainer = Trainer(profiler=profiler)

Reviewer 2: ```suggestion
    trainer = Trainer(profiler=HPUProfiler())
```
Or do I need to have a profiler instance to call summary?

    Self CPU time total: 1.681ms

.. note::
    Since HPUProfiler extends PyTorch Profiler, when using the HPUProfiler, wall clock time will not be representative of the true wall clock time.
    This is due to forcing profiled operations to be measured synchronously, when many HPU ops happen asynchronously.
    It is recommended to use this Profiler to find bottlenecks/breakdowns, however for end to end wall clock time use
    the ``SimpleProfiler``.

----

***************************
Profile a distributed model
***************************
To profile a distributed model, use the :class:`~pytorch_lightning.profilers.hpu.HPUProfiler` with the *filename* argument which will save a report per rank.

.. code-block:: python

    from pytorch_lightning.profilers import HPUProfiler

    profiler = HPUProfiler(filename="perf-logs")
    trainer = Trainer(profiler=profiler)

----

*****************************
Visualize profiled operations
*****************************
To visualize the profiled operations, enable **export_to_chrome** in the :class:`~pytorch_lightning.profilers.hpu.HPUProfiler` (Default: True).

.. code-block:: python

    from pytorch_lightning.profilers import HPUProfiler

>     profiler = HPUProfiler()

Reviewer 3: I know it's a default value for the moment, but then the code block doesn't really show anything.

```suggestion
    profiler = HPUProfiler(export_to_chrome=True)
```

4: View the profiling logs
==========================
Once the capture is finished, the page will refresh and you can browse through the insights using the **Tools** dropdown at the top left

################################################

> Find bottlenecks in your code on HPU (advanced)

Reviewer 4: @Felonious-Spellfire For review of docs structure here :)

    This is due to forcing profiled operations to be measured synchronously, when many HPU ops happen asynchronously.
    It is recommended to use this Profiler to find bottlenecks/breakdowns, however for end to end wall clock time use
    the ``SimpleProfiler``.

----

***************************
Profile a distributed model
***************************
To profile a distributed model, use the :class:`~pytorch_lightning.profilers.hpu.HPUProfiler` with the *filename* argument which will save a report per rank.

.. code-block:: python

    from pytorch_lightning.profilers import HPUProfiler

    profiler = HPUProfiler(filename="perf-logs")
    trainer = Trainer(profiler=profiler)

.. note::
    Since HPUProfiler extends PyTorch Profiler, when using the HPUProfiler, wall clock time will not not be representative of the true wall clock time.
    This is due to forcing profiled operations to be measured synchronously, when many HPU ops happen asynchronously.
    It is recommended to use this Profiler to find bottlenecks/breakdowns, however for end to end wall clock time use

>     the ``SimpleProfiler``.

Reviewer 5: The note is repeated.

Author: updated
